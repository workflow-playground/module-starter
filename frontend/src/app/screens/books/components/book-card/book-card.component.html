<li class="col-span-1 flex flex-col divide-y divide-gray-200 rounded-lg bg-white text-center shadow min-h-[510px]">
  <div class="flex flex-1 flex-col p-8">
    <div class="mx-auto flex-shrink-0" *ngIf="thumbnail">
      <ds-image [height]="185" [src]="thumbnail"></ds-image>
    </div>
    <h3 class="mt-6 text-base font-bold text-gray-900">{{ title | bcEllipsis: 35 }}</h3>

    <dl class="mt-1 flex flex-grow flex-col justify-between">
      <dd class="text-sm text-gray-500">{{ description | bcEllipsis }}</dd>
      <dd class="mt-3" *ngIf="authors.length > 0">
        <span
          *ngFor="let author of authors"
          class="inline-flex items-center rounded-full bg-primary-p300-main m-1 px-2 py-1 text-xs font-medium text-primary-p100 ring-1 ring-inset ring-primary-p300"
          >{{ author }}</span
        >
      </dd>
    </dl>
  </div>
  <ng-content></ng-content>
</li>
