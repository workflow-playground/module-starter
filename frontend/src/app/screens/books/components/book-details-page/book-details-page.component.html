<ds-book-details [book]="(book$ | async)!">
  <button
    (click)="onClick()"
    class="w-8 h-8 flex items-center justify-center rounded shadow"
    [ngClass]="{
      'bg-primary-p300-main': (isSelectedBookInCollection$ | async) === false,
      'bg-red-700': (isSelectedBookInCollection$ | async)
    }"
  >
    <ds-icon [size]="20" [name]="icons.Save" [color]="iconColors.white"></ds-icon>
  </button>
</ds-book-details>
