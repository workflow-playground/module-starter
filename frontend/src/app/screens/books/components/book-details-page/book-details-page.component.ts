import { ChangeDetectionStrategy, Component } from '@angular/core';
import { IconColor, IconName } from '@ds24/elements';

import { Store } from '@ngrx/store';
import { Observable, combineLatest, take } from 'rxjs';

import * as actions from '../../../../store/actions';
import * as selectors from '../../../../store/selectors';
import { Book } from '../../../../types/book.types';

@Component({
  selector: 'ds-book-details-page',
  changeDetection: ChangeDetectionStrategy.OnPush,
  templateUrl: './book-details-page.component.html',
})
export class BookDetailsPageComponent {
  icons: typeof IconName = IconName;
  iconColors: typeof IconColor = IconColor;

  book$ = this.store.select(selectors.books.selectSelectedBook) as Observable<Book>;
  isSelectedBookInCollection$ = this.store.select(selectors.collection.isSelectedBookInCollection);

  constructor(private store: Store) {}

  onClick() {
    combineLatest([this.book$, this.isSelectedBookInCollection$])
      .pipe(take(1))
      .subscribe(([book, isSelectedBookInCollection]) => {
        if (!isSelectedBookInCollection) {
          this.store.dispatch(actions.books.addBook({ book }));
        } else {
          this.store.dispatch(actions.books.removeBook({ book }));
        }
      });
  }
}
