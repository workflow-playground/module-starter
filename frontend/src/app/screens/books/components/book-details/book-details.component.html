<div class="w-full relative divide-gray-200 rounded-lg bg-white text-center shadow p-4">
  <div class="flex mb-5">
    <div class="mr-5 h-full" *ngIf="thumbnail">
      <ds-image [width]="150" [src]="thumbnail"></ds-image>
    </div>

    <div class="text-left">
      <h1 class="text-2xl text-neutral-n500 leading-tight">{{ title }}</h1>
      <h2 class="text-xl text-neutral-n300 p-0 mb-2 leading-tight" *ngIf="subtitle">{{ subtitle }}</h2>

      <div class="flex items-center mb-2" *ngIf="authors.length > 0">
        <span class="text-neutral-n300 mr-2">{{ 'BOOKS.DETAILS.AUTHORS' | translate }}</span>
        <span
          *ngFor="let author of authors"
          class="inline-flex items-center rounded-full bg-primary-p300-main px-2 py-1 m-1 text-xs font-medium text-primary-p100 ring-1 ring-inset ring-primary-p300"
          >{{ author }}</span
        >
      </div>

      <div class="flex items-center mb-2" *ngIf="published">
        <span class="text-neutral-n300 mr-2">{{ 'BOOKS.DETAILS.PUBLISHED' | translate }}</span>
        <span class="text-neutral-n500">{{ published }}</span>
      </div>

      <div class="flex items-center mb-2" *ngIf="language">
        <span class="text-neutral-n300 mr-2">{{ 'BOOKS.DETAILS.LANGUAGE' | translate }}</span>
        <span class="text-neutral-n500">{{ language }}</span>
      </div>
    </div>
  </div>
  <blockquote class="text-left italic text-neutral-n500" [innerHtml]="description"></blockquote>
  <div class="absolute top-4 right-4">
    <ng-content></ng-content>
  </div>
</div>
