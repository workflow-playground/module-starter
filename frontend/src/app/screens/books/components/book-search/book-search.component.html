<div class="relative rounded-md shadow-sm mb-5">
  <input
    type="text"
    [value]="query"
    (keyup)="onSearch($event)"
    class="block text-base w-full rounded-md border-0 py-2 pl-2 pr-10 text-neutral-n500 ring-1 ring-inset ring-neutral-n200 placeholder:text-base placeholder:text-neutral-n300 focus:ring-2 focus:ring-inset focus:ring-primary-p300-main sm:text-sm sm:leading-6"
    [placeholder]="'BOOKS.SEARCH.PLACEHOLDER' | translate"
  />
  <div class="pointer-events-none absolute inset-y-0 right-0 flex items-center pr-3">
    <ds-icon *ngIf="!searching" [name]="icons.MagnifyingGlass" [size]="20" [color]="iconColors.Neutral200"></ds-icon>
    <ds-loader *ngIf="searching" size="small"></ds-loader>
  </div>
</div>
<div class="text-danger-e500" *ngIf="error">{{ error }}</div>
