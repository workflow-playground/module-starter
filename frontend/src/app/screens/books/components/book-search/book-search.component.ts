import { Component, Output, Input, EventEmitter, ChangeDetectionStrategy } from '@angular/core';
import { IconColor, IconName } from '@ds24/elements';

@Component({
  selector: 'ds-book-search',
  templateUrl: './book-search.component.html',
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class BookSearchComponent {
  @Input() query = '';
  @Input() searching = false;
  @Input() error = '';
  @Output() search = new EventEmitter<string>();

  icons: typeof IconName = IconName;
  iconColors: typeof IconColor = IconColor;

  onSearch(event: KeyboardEvent): void {
    this.search.emit((event.target as HTMLInputElement).value);
  }
}
