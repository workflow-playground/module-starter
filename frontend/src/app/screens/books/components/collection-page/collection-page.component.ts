import { ChangeDetectionStrategy, Component } from '@angular/core';
import { Store } from '@ngrx/store';
import { IconColor, IconName } from '@ds24/elements';

import * as selectors from '../../../../store/selectors';

@Component({
  selector: 'ds-collection-page',
  changeDetection: ChangeDetectionStrategy.OnPush,
  templateUrl: './collection-page.component.html',
})
export class CollectionPageComponent {
  icons: typeof IconName = IconName;
  iconColors: typeof IconColor = IconColor;

  books$ = this.store.select(selectors.collection.selectBookCollection);

  constructor(private store: Store) {}
}
