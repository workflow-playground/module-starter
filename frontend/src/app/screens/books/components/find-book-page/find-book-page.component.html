<ds-book-search
[query]="(searchQuery$ | async)!"
[searching]="(loading$ | async)!"
[error]="(error$ | async)!"
(search)="search($event)"
>
</ds-book-search>
<ul role="list" class="grid grid-cols-1 gap-6 sm:grid-cols-2 md:grid-cols-3">
<ds-book-card *ngFor="let book of books$ | async" [book]="book">
  <div class="-mt-px flex divide-x divide-gray-200">
    <div class="flex w-0 flex-1">
      <a
        [routerLink]="['/books', book.id]"
        class="relative -mr-px inline-flex w-0 flex-1 items-center justify-center gap-x-3 rounded-bl-lg border border-transparent py-4 text-sm font-bold text-neutral-n500 hover:text-primary-p500 overflow-hidden"
      >
        <ds-icon [name]="icons.File" [size]="20" [color]="iconColors.Primary400"></ds-icon>
        {{ 'BOOKS.COLLECTION.SHOW_DETAILS' | translate }}
      </a>
    </div>
  </div>
</ds-book-card>
</ul>
