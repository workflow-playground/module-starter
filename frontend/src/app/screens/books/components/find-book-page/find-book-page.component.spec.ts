import { ComponentFixture, TestBed } from '@angular/core/testing';
import { ReactiveFormsModule } from '@angular/forms';
import { IconModule, ImageModule, LoaderModule } from '@ds24/elements';

import { TranslateModule } from '@ngx-translate/core';
import { MockStore, provideMockStore } from '@ngrx/store/testing';

import { FindBookPageComponent } from './find-book-page.component';
import { BookDetailsComponent } from '../book-details/book-details.component';
import { BookSearchComponent } from '../book-search/book-search.component';
import { BookCardComponent } from '../book-card/book-card.component';
import { CollectionPageComponent } from '../collection-page/collection-page.component';
import { BookDetailsPageComponent } from '../book-details-page/book-details-page.component';

import { State, actions, selectors } from '../../../../store';

describe('Find Book Page', () => {
  let fixture: ComponentFixture<FindBookPageComponent>;
  let mockStore: MockStore<State>;
  let instance: FindBookPageComponent;

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [IconModule, ReactiveFormsModule, TranslateModule.forRoot(), ImageModule, LoaderModule],
      declarations: [
        BookDetailsComponent,
        BookSearchComponent,
        BookCardComponent,
        CollectionPageComponent,
        BookDetailsPageComponent,
        FindBookPageComponent,
      ],
      providers: [provideMockStore()],
    });

    fixture = TestBed.createComponent(FindBookPageComponent);
    instance = fixture.componentInstance;
    mockStore = TestBed.inject(MockStore);

    mockStore.overrideSelector(selectors.search.selectSearchQuery, '');
    mockStore.overrideSelector(selectors.books.selectSearchResults, []);
    mockStore.overrideSelector(selectors.search.selectSearchLoading, false);
    mockStore.overrideSelector(selectors.search.selectSearchError, '');

    jest.spyOn(mockStore, 'dispatch');
  });

  it('should compile', () => {
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should dispatch a book.Search action on search', () => {
    const $event = 'book name';
    const action = actions.search.searchBooks({ query: $event });

    instance.search($event);

    expect(mockStore.dispatch).toHaveBeenCalledWith(action);
  });
});
