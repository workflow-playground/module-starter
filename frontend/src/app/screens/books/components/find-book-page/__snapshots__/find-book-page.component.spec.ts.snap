// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Find Book Page should compile 1`] = `
<ds-find-book-page
  books$={[Function _Store]}
  error$={[Function _Store]}
  iconColors={[Function Object]}
  icons={[Function Object]}
  loading$={[Function _Store]}
  searchQuery$={[Function _Store]}
  store={[Function _MockStore]}
>
  <ds-book-search>
    <div
      class="relative rounded-md shadow-sm mb-5"
    >
      <input
        class="block text-base w-full rounded-md border-0 py-2 pl-2 pr-10 text-neutral-n500 ring-1 ring-inset ring-neutral-n200 placeholder:text-base placeholder:text-neutral-n300 focus:ring-2 focus:ring-inset focus:ring-primary-p300-main sm:text-sm sm:leading-6"
        placeholder="BOOKS.SEARCH.PLACEHOLDER"
        type="text"
      />
      <div
        class="pointer-events-none absolute inset-y-0 right-0 flex items-center pr-3"
      >
        <ds-icon>
          <magnifying-glass>
            <svg
              fill="none"
              height="20px"
              style="height: 20px; width: 20px;"
              viewBox="0 0 40 40"
              width="20px"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                class="fill-neutral-n200"
                clip-rule="evenodd"
                d="M8.29142 27.8438C3.21107 25.4221 -0.0176564 20.2883 0.000121082 14.6603C-0.0259332 8.2324 4.15507 2.54368 10.2976 0.64945C16.4401 -1.24478 23.0988 1.10118 26.6975 6.42736C30.2962 11.7535 29.988 18.8067 25.9386 23.7988L39.5694 37.368L37.354 39.5834L23.754 25.9834C19.3931 29.5411 13.3718 30.2656 8.29142 27.8438ZM14.6155 26.1988C20.988 26.1988 26.154 21.0328 26.154 14.6603C26.1622 11.5976 24.9491 8.65798 22.7835 6.49232C20.6178 4.32666 17.6782 3.11364 14.6155 3.12184C8.24299 3.12184 3.07704 8.28778 3.07704 14.6603C3.07704 21.0328 8.24299 26.1988 14.6155 26.1988Z"
                fill-rule="evenodd"
              />
            </svg>
          </magnifying-glass>
        </ds-icon>
      </div>
    </div>
  </ds-book-search><ul
    class="grid grid-cols-1 gap-6 sm:grid-cols-2 md:grid-cols-3"
    role="list"
  >
  </ul>
</ds-find-book-page>
`;
